File: SWEA/5658.cpp
#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;
bool comp(int a, int b)
{
  return a > b;
}
int main(int argc, char **argv)
{
  vector<int> answers;
  int test_case;
  int T;
  cin >> T;
  
  for (test_case = 1; test_case <= T; ++test_case)
  {
    int n, k, d;
    string nums;
    cin >> n >> k;
    cin >> nums;
    d = n / 4;
    vector<int> result;
    for(int i = 0; i < d+1; i++){
      for(int j = 0; j < d+1; j++){
        int start = d * j;
        int end = d * (j + 1);

        string tmp = nums.substr(start, end);
        int a = strtol(tmp.c_str(), NULL, 16);

        result.push_back(a);
      }
      char ls = nums[n - 1];
      nums.erase(n - 1);
      nums.insert(0, ls);

    }
    sort(result.begin(), result.end());
    result.erase(unique(result.begin(), result.end()), result.end());
    sort(result.begin(), result.end(), comp);
    
    answers.push_back(result[k - 1]);
    
    for(auto it=result.begin(); it != result.end(); it++){
      cout << *it << " ";
    }
  }

  for(int i = 0; i < answers.size(); i++){
    cout << "#" << i+1 << " " << answers[i] << endl;
  }

  return 0;
}

File: SWEA/5658.py
# 보물상자 비밀번호
# https://swexpertacademy.com/main/code/problem/problemDetail.do?contestProbId=AWXRUN9KfZ8DFAUo&categoryId=AWXRUN9KfZ8DFAUo&categoryType=CODE&problemTitle=%EB%B3%B4%EB%AC%BC&orderBy=FIRST_REG_DATETIME&selectCodeLang=ALL&select-1=&pageSize=10&pageIndex=1

t = int(input())
answers = []
for _ in range(t):
    n, k = map(int, input().split())
    d = n // 4 # 묶이는 개수

    nums = input()
    arr = []
    
    for num in nums:
        arr.append(num)
        # num = int(num, 16)
    
    result = []

    tmp = ""
    i = 0

    for tt in range(d+1):
        for i in range(4):
            start = d * i
            end = d * (i + 1)
            # result.append(int(''.join(arr[start:end]), 16))
            result.append(''.join(arr[start:end]))
        temp = arr.pop()
        arr.insert(0, temp)

    result = list(set(result))
    result.sort(reverse=True)
    answers.append(result[k-1])

for i in range(len(answers)):
    print(f"#{i+1} {int(answers[i], 16)}")
